services/diagnostic/handlers.go

```diff
@@ -0,0 +1,1222 @@
+package diagnostic
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"log"
+	"runtime"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/influxdata/kapacitor"
+	"github.com/influxdata/kapacitor/alert"
+	"github.com/influxdata/kapacitor/edge"
+	"github.com/influxdata/kapacitor/keyvalue"
+	"github.com/influxdata/kapacitor/models"
+	alertservice "github.com/influxdata/kapacitor/services/alert"
+	"github.com/influxdata/kapacitor/services/alerta"
+	"github.com/influxdata/kapacitor/services/hipchat"
+	"github.com/influxdata/kapacitor/services/httppost"
+	"github.com/influxdata/kapacitor/services/influxdb"
+	"github.com/influxdata/kapacitor/services/k8s"
+	"github.com/influxdata/kapacitor/services/mqtt"
+	"github.com/influxdata/kapacitor/services/opsgenie"
+	"github.com/influxdata/kapacitor/services/pagerduty"
+	"github.com/influxdata/kapacitor/services/pushover"
+	"github.com/influxdata/kapacitor/services/sensu"
+	"github.com/influxdata/kapacitor/services/slack"
+	"github.com/influxdata/kapacitor/services/smtp"
+	"github.com/influxdata/kapacitor/services/snmptrap"
+	"github.com/influxdata/kapacitor/services/swarm"
+	"github.com/influxdata/kapacitor/services/talk"
+	"github.com/influxdata/kapacitor/services/telegram"
+	"github.com/influxdata/kapacitor/services/udp"
+	"github.com/influxdata/kapacitor/services/victorops"
+	plog "github.com/prometheus/common/log"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+func Error(l *zap.Logger, msg string, err error, ctx []keyvalue.T) {
+	if len(ctx) == 0 {
+		l.Error(msg, zap.Error(err))
+		return
+	}
+
+	if len(ctx) == 1 {
+		el := ctx[0]
+		l.Error(msg, zap.Error(err), zap.String(el.Key, el.Value))
+		return
+	}
+
+	if len(ctx) == 2 {
+		x := ctx[0]
+		y := ctx[1]
+		l.Error(msg, zap.Error(err), zap.String(x.Key, x.Value), zap.String(y.Key, y.Value))
+		return
+	}
+
+	// This isn't great wrt to allocation, but should not ever actually occur
+	fields := make([]zapcore.Field, len(ctx)+1) // +1 for error
+	fields[0] = zap.Error(err)
+	for i := 1; i < len(fields); i++ {
+		kv := ctx[i-1]
+		fields[i] = zap.String(kv.Key, kv.Value)
+	}
+
+	l.Error(msg, fields...)
+}
+
+// Alert Service Handler
+
+type AlertServiceHandler struct {
+	l *zap.Logger
+}
+
+func zapFieldsFromContext(ctx []keyvalue.T) []zapcore.Field {
```

Comment:
Why not use this above on lines 62-67?

Reply:
The first entry of the slice  on line 63 is `fields[0] = zap.Error(err)` is an error. This method expects that they all be strings.